### DISCUSSION TITLE
How do I work with OrtCUDAProviderOptionsV2 struct using c++ api header ?

### DISCUSSION THREAD

**Discussion by boxerab** [microsoft/onnxruntime#16627] [General] [answered]
I am on tag 1.15.1.  The code below compiles and runs, but since the c++ header doesn't define `OrtCUDAProviderOptionsV2`, I can't set any of the structure's fields. I would like to set the deviceId.

```
          OrtCUDAProviderOptionsV2 *cuda_options = nullptr;
          Ort::ThrowOnError (api.CreateCUDAProviderOptions (&cuda_options));
          std::unique_ptr < OrtCUDAProviderOptionsV2,
              decltype (api.ReleaseCUDAProviderOptions) >
              rel_cuda_options (cuda_options, api.ReleaseCUDAProviderOptions);
          Ort::
              ThrowOnError (api.SessionOptionsAppendExecutionProvider_CUDA_V2
              (static_cast < OrtSessionOptions * >(sessionOptions),
                  rel_cuda_options.get ()));
```

**Comment by baijumeswani** [comment, marked as answer]
This can be done using the function `UpdateCUDAProviderOptions`.

For example, you could do:

```cc
const std::vector<std::string> keys{"device_id"};
const std::vector<std::string> values{"1"};
api.UpdateCUDAProviderOptions(cuda_options, keys.data(), values.data(), 1U);
```

Refer to the API documentation [here](https://onnxruntime.ai/docs/api/c/struct_ort_api.html#a4710fc51f75a4b9a75bde20acbfa0783)

Note that the code snippet I provided is not tested and may need adjustments.

**Reply by boxerab** [reply]
@baijumeswani thank you !!! Will I eventually be able to simply set the field on the struct with later releases?

**Reply by baijumeswani** [reply]
It would seem unlikely considering that the type was made opaque by intention. Perhaps the reason it was made opaque to begin with was to hide the implementation and allow the devs to expand/change its definition without impacting backwards compatibility.

But I could be mistaken.

**Reply by boxerab** [reply]
Thanks, I can see the logic here ; not having to include any EP dependant headers in order to set the fields.
